File: v1/validation_builder.go
package smartform

// ValidationBuilder provides a fluent API for creating validation rules
type ValidationBuilder struct{}

// NewValidationBuilder creates a new validation builder
func NewValidationBuilder() *ValidationBuilder {
	return &ValidationBuilder{}
}

// Required creates a required validation rule
func (vb *ValidationBuilder) Required(message string) *ValidationRule {
	return &ValidationRule{
		Type:    ValidationTypeRequired,
		Message: message,
	}
}

func (vb *ValidationBuilder) RequiredIf(condition *Condition, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeRequiredIf,
		Message:    message,
		Parameters: condition,
	}
}

// MinLength creates a minimum length validation rule
func (vb *ValidationBuilder) MinLength(min float64, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeMinLength,
		Message:    message,
		Parameters: min,
	}
}

// MaxLength creates a maximum length validation rule
func (vb *ValidationBuilder) MaxLength(max float64, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeMaxLength,
		Message:    message,
		Parameters: max,
	}
}

// Pattern creates a pattern validation rule
func (vb *ValidationBuilder) Pattern(pattern string, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypePattern,
		Message:    message,
		Parameters: pattern,
	}
}

// Min creates a minimum value validation rule
func (vb *ValidationBuilder) Min(min float64, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeMin,
		Message:    message,
		Parameters: min,
	}
}

// Max creates a maximum value validation rule
func (vb *ValidationBuilder) Max(max float64, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeMax,
		Message:    message,
		Parameters: max,
	}
}

// Email creates an email validation rule
func (vb *ValidationBuilder) Email(message string) *ValidationRule {
	return &ValidationRule{
		Type:    ValidationTypeEmail,
		Message: message,
	}
}

// URL creates a URL validation rule
func (vb *ValidationBuilder) URL(message string) *ValidationRule {
	return &ValidationRule{
		Type:    ValidationTypeURL,
		Message: message,
	}
}

// FileType creates a file type validation rule
func (vb *ValidationBuilder) FileType(allowedTypes []string, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeFileType,
		Message:    message,
		Parameters: allowedTypes,
	}
}

// FileSize creates a file size validation rule
func (vb *ValidationBuilder) FileSize(maxSize float64, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeFileSize,
		Message:    message,
		Parameters: maxSize,
	}
}

// ImageDimensions creates an image dimensions validation rule
func (vb *ValidationBuilder) ImageDimensions(dimensions map[string]interface{}, message string) *ValidationRule {
	return &ValidationRule{
		Type:       ValidationTypeImageDimensions,
		Message:    message,
		Parameters: dimensions,
	}
}

// Dependency creates a field dependency validation rule
func (vb *ValidationBuilder) Dependency(field string, operator string, value interface{}, message string) *ValidationRule {
	return &ValidationRule{
		Type:    ValidationTypeDependency,
		Message: message,
		Parameters: map[string]interface{}{
			"field":    field,
			"operator": operator,
			"value":    value,
		},
	}
}

// Unique creates a uniqueness validation rule
func (vb *ValidationBuilder) Unique(message string) *ValidationRule {
	return &ValidationRule{
		Type:    ValidationTypeUnique,
		Message: message,
	}
}

// Custom creates a custom validation rule
func (vb *ValidationBuilder) Custom(functionName string, params map[string]interface{}, message string) *ValidationRule {
	if params == nil {
		params = make(map[string]interface{})
	}
	params["function"] = functionName

	return &ValidationRule{
		Type:       ValidationTypeCustom,
		Message:    message,
		Parameters: params,
	}
}

File: frontend/core/src/validation/validation-builder.ts
import { ValidationRule, ValidationType } from "../types";

/**
 * Builder for creating validation rules
 */
export class ValidationBuilder {
  /**
   * Create a required validation rule
   */
  public required(message: string): ValidationRule {
    return {
      type: ValidationType.REQUIRED,
      message,
    };
  }

  /**
   * Create a minimum length validation rule
   */
  public minLength(min: number, message: string): ValidationRule {
    return {
      type: ValidationType.MIN_LENGTH,
      message,
      parameters: min,
    };
  }

  /**
   * Create a maximum length validation rule
   */
  public maxLength(max: number, message: string): ValidationRule {
    return {
      type: ValidationType.MAX_LENGTH,
      message,
      parameters: max,
    };
  }

  /**
   * Create a pattern validation rule
   */
  public pattern(pattern: string, message: string): ValidationRule {
    return {
      type: ValidationType.PATTERN,
      message,
      parameters: pattern,
    };
  }

  /**
   * Create a minimum value validation rule
   */
  public min(min: number, message: string): ValidationRule {
    return {
      type: ValidationType.MIN,
      message,
      parameters: min,
    };
  }

  /**
   * Create a maximum value validation rule
   */
  public max(max: number, message: string): ValidationRule {
    return {
      type: ValidationType.MAX,
      message,
      parameters: max,
    };
  }

  /**
   * Create an email validation rule
   */
  public email(message: string): ValidationRule {
    return {
      type: ValidationType.EMAIL,
      message,
    };
  }

  /**
   * Create a URL validation rule
   */
  public url(message: string): ValidationRule {
    return {
      type: ValidationType.URL,
      message,
    };
  }

  /**
   * Create a file type validation rule
   */
  public fileType(allowedTypes: string[], message: string): ValidationRule {
    return {
      type: ValidationType.FILE_TYPE,
      message,
      parameters: allowedTypes,
    };
  }

  /**
   * Create a file size validation rule
   */
  public fileSize(maxSize: number, message: string): ValidationRule {
    return {
      type: ValidationType.FILE_SIZE,
      message,
      parameters: maxSize,
    };
  }

  /**
   * Create an image dimensions validation rule
   */
  public imageDimensions(
    dimensions: Record<string, any>,
    message: string,
  ): ValidationRule {
    return {
      type: ValidationType.IMAGE_DIMENSIONS,
      message,
      parameters: dimensions,
    };
  }

  /**
   * Create a field dependency validation rule
   */
  public dependency(
    field: string,
    operator: string,
    value: any,
    message: string,
  ): ValidationRule {
    return {
      type: ValidationType.DEPENDENCY,
      message,
      parameters: {
        field,
        operator,
        value,
      },
    };
  }

  /**
   * Create a uniqueness validation rule
   */
  public unique(message: string): ValidationRule {
    return {
      type: ValidationType.UNIQUE,
      message,
    };
  }

  /**
   * Create a custom validation rule
   */
  public custom(
    functionName: string,
    params: Record<string, any>,
    message: string,
  ): ValidationRule {
    if (params === null || params === undefined) {
      params = {};
    }
    params.function = functionName;

    return {
      type: ValidationType.CUSTOM,
      message,
      parameters: params,
    };
  }
}

/**
 * Create a new validation builder
 */
export function createValidationBuilder(): ValidationBuilder {
  return new ValidationBuilder();
}
